butte-build/src/codegen.rs

```diff
@@ -190,14 +166,42 @@ fn to_type_token(
     }
 }
 
+/// A trait to format literals according to a provided type.
+trait PrimitiveLiteralToken {
+    /// The type of the token that will be returned formatted according to the provided type.
+    type TokenType;
+
+    fn fmt_lit(&self, ty: impl Spanned + fmt::Display) -> Self::TokenType;
+}
+
+impl PrimitiveLiteralToken for ast::IntegerConstant {
+    type TokenType = syn::LitInt;
+
+    fn fmt_lit(&self, ty: impl Spanned + fmt::Display) -> Self::TokenType {
+        Self::TokenType::new(&format!("{}{}", self, ty), ty.span())
+    }
+}
+
+impl PrimitiveLiteralToken for ast::FloatingConstant {
+    type TokenType = syn::LitFloat;
+
+    fn fmt_lit(&self, ty: impl Spanned + fmt::Display) -> Self::TokenType {
+        Self::TokenType::new(&format!("{}{}", self, ty), ty.span())
```

Comment:
The tests fail because it doesn't truncate the 64bit `FloatingConstant` to the target type.

The following unbreaks the tests, is there a better way to do this?

```suggestion
        let float_ty = format!("{}", ty);
        match float_ty.as_ref() {
            "f32" => Self::TokenType::new(&format!("{}{}", *self as f32, float_ty), ty.span()),
            _ => Self::TokenType::new(&format!("{}{}", self, float_ty), ty.span()),
        }
```

Reply:
Yeah I had noticed this as well, but didn't a chance to come back to this. The fix looks ok if there's no better way to get the type out. I'll take a look for a bit and if nothing emerges I'll use this fix.

Reply:
Alright, I ended up going with this fix (slightly rewritten, but semantically equivalent).

Reply:
Thanks for this @reyk!